packages/ecs/src/QueryFunctions.test.tsx

```diff
@@ -0,0 +1,181 @@
+  assert.doesNotThrow(Q, err)
+}
+
+function assertQueryNotOk(Q: Query, err = 'Query is Ok') {
+  assert.throws(Q.enter, err)
+  assert.throws(Q.exit, err)
+  assert.throws(Q, err)
+}
+
+function assertDefinedQuery(Q: Query, expected: Entity[]) {
+  assertQueryOk(Q, 'The test query did not get defined correctly')
+  assertArrayNotEqual(Q(), [], 'The query did not return any entities.')
+  assertArrayEqual(Q(), expected, 'The test query did not return the expected result')
+}
+
+describe('QueryFunctions', () => {
+  const component = defineComponent({ name: 'TestComponent' })
+  let entity1 = UndefinedEntity
+  let entity2 = UndefinedEntity
+
+  beforeEach(() => {
+    startEngine()
+    ComponentMap.clear()
+    entity1 = createEntity()
+    entity2 = createEntity()
+  })
+
+  afterEach(() => {
+    removeEntity(entity1)
+    removeEntity(entity2)
+    return destroyEngine()
+  })
+
+  describe('defineQuery', () => {
+    it('should create a valid query', () => {
+      const query = defineQuery([component])
+      setComponent(entity1, component)
+      setComponent(entity2, component)
+      assertDefinedQuery(query, [entity1, entity2])
+    })
+  })
+
+  /**
+   *  @todo : Why does `query()` still work after calling `removeQuery(query)`
+   *  @todo : Why does `assertQueryNotOk` not work? Is it because of the above?
+   */
+  //  describe('removeQuery', () => {
+  //   it('should remove a query when called', () => {
+  //     const query = defineQuery([component])
+  //     setComponent(entity1, component)
+  //     setComponent(entity2, component)
+  //     assertDefinedQuery(query, [entity1, entity2])
+  //
+  //     removeQuery(query)
+  //     console.log(" --->", query())
+  //     assertQueryNotOk(query, "The test query did not get removed correctly")
+  //   })
+  // })
+  //
+})
+
+describe('QueryFunctions Hooks', async () => {
+  describe('useQuery', async () => {
+    const component = defineComponent({ name: 'TestComponent' })
+    const InitialValue = [123456789 as Entity, 987654321 as Entity]
+    let entity1 = UndefinedEntity
+    let entity2 = UndefinedEntity
+    let result = InitialValue
+
+    beforeEach(() => {
+      startEngine()
+      ComponentMap.clear()
+      entity1 = createEntity()
+      entity2 = createEntity()
+    })
+
+    afterEach(() => {
+      removeEntity(entity1)
+      removeEntity(entity2)
+      return destroyEngine()
+    })
+
+    // Define the Reactor that will run the tested hook
+    const Reactor = () => {
+      const data = useQuery([component])
+      console.log('render', data)
+      useEffect(() => {
+        console.log('effect', data)
+        result = data as Entity[]
+      }, [data])
+      return null
+    }
+
+    it('should return an empty array when entities have the component', async () => {
+      const ExpectedValue = [] as Entity[]
+      const tag = <Reactor />
+      const { rerender, unmount } = render(tag)
+      await act(() => rerender(tag))
+      assertArrayNotEqual(
```

Comment:
i'm not a fan of asserting against some arbitrary value. we should just be asserting the data we expect. this is functionally equivalent as if `result` is undefined

Reply:
we should be asserting both that the value is undefined and that the useEffect has run. something like a counter increment along with the result data would suffice